### PR TITLE
Ed shel/features/registerAndAuth

### PR DESCRIPTION
**Pull request by EdShel**
Working registration/authentication using JWT tokens and Identity configured in appsettings.json.

### CODE REVIEW

File: UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        private async Task<ClaimsIdentity> GetIdentity(LoginRequest loginModel)
+        {
+            User user = await userManager.FindByEmailAsync(loginModel.Email);
+            if (user != null)
+            {
+                bool passwordIsCorrect = await userManager.CheckPasswordAsync(user, loginModel.Password);
+                if (passwordIsCorrect)
+                {
+                    var userClaims = await userManager.GetClaimsAsync(user);
+                    userClaims.Add(new Claim("UserId<==", user.Id));
```
Comment (EdShel) [inline UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: Just passing User id with an extraordinary name in order to work with it when will be making authorization. Consider this as TODO.

File: UniAtHome/UniAtHome.WebAPI/appsettings.Development.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "DefaultConnection": "Your local db here"
+    "DefaultConnection": "Data Source=(localdb)\\MSSQLLocalDB;Initial Catalog=UniAtHomeDatabase;Integrated Security=True;Connect Timeout=30;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False"
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/appsettings.Development.json:3]: Please, don't change connection string  here

File: UniAtHome/UniAtHome.WebAPI/appsettings.json [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
       "Microsoft": "Warning",
       "Microsoft.Hosting.Lifetime": "Information"
     }
   },
-  "AllowedHosts": "*"
+  "AllowedHosts": "*",
+  "AuthenticationToken": {
+    "ValidIssuer": "UniAtHomeBackEnd",
+    "ValidAudience": "UniAtHomeFrontEndUser",
+    "SigningKey": "abra-kadabra-must-be-long-sobaka-dog",
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/appsettings.json:16]: Great key!

File: UniAtHome/UniAtHome.WebAPI/appsettings.json [hunk 4]
```diff
@@ -1,10 +1,20 @@
 {
+  "ConnectionStrings": {
+    "DefaultConnection": "Data Source=(localdb)\\MSSQLLocalDB;Initial Catalog=UniAtHomeDatabase;Integrated Security=True;Connect Timeout=30;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False"
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/appsettings.json:3]: And here too.

File: UniAtHome/UniAtHome.WebAPI/Extensions/StartupExtensions.cs [hunk 5]
```diff
@@ -20,7 +45,11 @@ public static IServiceCollection AddIdentityConfiguration(this IServiceCollectio
             services.AddIdentity<User, IdentityRole>(
                     config =>
                     {
-                        //Please, add identity configuration here
+                        // Our UserNames = Emails. When registering new user with already registered email
+                        // there will be two unique violations unless emails ain't requiered to be unique.
+                        config.User.RequireUniqueEmail = false;
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/Extensions/StartupExtensions.cs:57]: I guess we should change it to "true"

File: UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        public async Task<object> Register([FromBody] RegistrationRequest registerModel)
+        {
+            var user = new User
+            {
+                UserName = registerModel.Email,
+                FirstName = registerModel.FirstName,
+                LastName = registerModel.LastName,
+                Email = registerModel.Email
+            };
+            var registerResult = await userManager.CreateAsync(user, registerModel.Password);
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: Please, don't use UserManager here. We should use Services and Repositories.

File: UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            this.userManager = userManager;
+            this.tokenGenerator = tokenGenerator;
+        }
+
+        [AllowAnonymous]
+        [HttpPost("register")]
+        public async Task<object> Register([FromBody] RegistrationRequest registerModel)
+        {
+            var user = new User
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: Please, don't create User instance here. We shouldn't use Entity classes in Controllers. Create DTO object i you need it.
In future wi will use AutoMapper

File: UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            var registerResult = await userManager.CreateAsync(user, registerModel.Password);
+            if (registerResult.Succeeded)
+            {
+                return Ok("User is successfully registered!");
+            }
+
+            return GetErrorOfNotSucceededRequest(registerResult);
+        }
+
+        private ObjectResult GetErrorOfNotSucceededRequest(IdentityResult registerResult)
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: Why do you use ObjectResult instead ActionResult? What diference between them?

Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: Not sure that this method is the best solution.

File: UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    [ApiController]
+    public class UsersController : ControllerBase
+    {
+        private readonly UserManager<User> userManager;
+
+        private readonly IAuthTokenGenerator tokenGenerator;
+
+        public UsersController(UserManager<User> userManager, IAuthTokenGenerator tokenGenerator)
+        {
+            this.userManager = userManager;
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: User manager shouldn't be used in Controllers

File: UniAtHome/UniAtHome.DAL/UniAtHomeDbContext.cs [hunk 10]
```diff
@@ -1,20 +1,20 @@
...
-    public class UniAtHomeDbContext : DbContext
+    public class UniAtHomeDbContext : IdentityDbContext<User>
     {
         public UniAtHomeDbContext(DbContextOptions<UniAtHomeDbContext> options) : base(options)
         {
         }
 
         public DbSet<Course> Courses { get; set; }
         public DbSet<Lesson> Lessons { get; set; }
-        public DbSet<User> Users { get; set; }
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.DAL/UniAtHomeDbContext.cs:17]: Why did you remove this?

File: UniAtHome/UniAtHome.DAL/Migrations/UniAtHomeDbContextModelSnapshot.cs [hunk 11]
```diff
@@ -19,6 +19,137 @@ protected override void BuildModel(ModelBuilder modelBuilder)
                 .HasAnnotation("Relational:MaxIdentifierLength", 128)
                 .HasAnnotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn);
 
+            modelBuilder.Entity("Microsoft.AspNetCore.Identity.IdentityRole", b =>
```
Comment (arstrybuh257) [inline UniAtHome/UniAtHome.DAL/Migrations/UniAtHomeDbContextModelSnapshot.cs:22]: Please, don't modify autogenerated tables

**Review by arstrybuh257 — CHANGES_REQUESTED** [review]

File: UniAtHome/UniAtHome.WebAPI/Extensions/StartupExtensions.cs (hunk 5 above)
Comment (EdShel) [reply UniAtHome/UniAtHome.WebAPI/Extensions/StartupExtensions.cs:57]: If it's true, the default error message about registering 2 users with the same email will produce 2 errors of the same kind ( 1) already exists user with this name 2) already exists user with this email). So when this property is false, there is only message #1.

File: UniAtHome/UniAtHome.DAL/UniAtHomeDbContext.cs (hunk 10 above)
Comment (EdShel) [reply UniAtHome/UniAtHome.DAL/UniAtHomeDbContext.cs:17]: Context now extends identity context, so it has DbSet<User> inherited from the parent

File: UniAtHome/UniAtHome.DAL/Migrations/UniAtHomeDbContextModelSnapshot.cs (hunk 11 above)
Comment (EdShel) [reply UniAtHome/UniAtHome.DAL/Migrations/UniAtHomeDbContextModelSnapshot.cs:22]: It has automagically updated after Migration-Add. I don't even know what it is :P

File: UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs (hunk 8 above)
Comment (EdShel) [reply UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: I thought ObjectResult takes into account user's response preferences (XML, Json) and serializes returned object this way.

File: UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs (hunk 6 above)
Comment (EdShel) [reply UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: Shall I encapsulate UserManager inside UsersRepository? However, in order to create the Repository I'll need UserManager dependency. There is no way to remove it.
Or do I need to recreate the whole UserManager functions from scratch including password verification, adding roles and claims etc?
Or can I extend UserManager by Repository?

File: UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs (hunk 7 above)
Comment (EdShel) [reply UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: UserManager works specifically with User objects as well the UserRepository does. Shall I remove all the code and put it inside BLL?

Comment (lesloress) [reply UniAtHome/UniAtHome.WebAPI/Controllers/UsersController.cs]: All business logic should be in bll. Managing users refers to business logic.

**Review by arstrybuh257 — APPROVED** [review]
